### ISSUE TITLE
Patient search 

### ISSUE THREAD

**Issue by f-odhiambo** [opensrp/fhircore#20]
User story :
As a CHW, I want to do a basic search for  a patient so that I can be able to retrieve their details 

Acceptance criteria :
- [x] Filter by First name or Filter by the Last name
- [x] Display results
- [x] On clicking a patient show patient [profile view](https://github.com/OpenSRP/fhircore/issues/22) 

This is a subtask of https://github.com/OpenSRP/fhircore/issues/50

**Comment by f-odhiambo** [comment]
User story :
As a CHW, I want to search for  a patient so that I can be able to retrieve their details 

- Search can be implemented based on FHIR [parameters](https://www.hl7.org/fhir/search.html) 
- Search will be based on the search API in the  Android FHIR SDK 

Acceptance criteria :
- [ ] Clicking on the search icon should present you with a patient search screen 
- [ ] Able to add the search criteria via a selection list 
- [ ] If the query returns a result,  display  search results and on select view and on click view patient details in a [patient profile view](https://github.com/OpenSRP/fhircore/issues/22) 
- [ ] If the query is  null result display not found message/notification

**Comment by vincent-karuri** [comment]
@f-odhiambo Is it possible to separate out the portions into sub-issues to bring better visibility to what portion is being worked on by different people?

From the call with Google, it seems like we have some portion of search covered by the SDK already (for full-name matches). The search functionality here seems to be more advanced and I think there's room to incrementally add support for each feature e.g. initially we can support some subset of parameters, then we can add boolean logic (OR, AND) etc.

**Comment by jingtang10** [comment]
https://github.com/google/android-fhir/issues/57

**Comment by pld** [comment]
@rkodev @f-odhiambo in-line w/ @vincent-karuri 's question, can you break down what will be addressed here in fhircore vs in [android-fhir](https://github.com/google/android-fhir)?

**Comment by rkodev** [comment]
@jingtang10 @pld @f-odhiambo 
At the SDK level, we can have the features
1. [Contains Modifier](https://www.hl7.org/fhir/search.html#modifiers) Partial search using modifiers enabled.
2. Allowing any set of filter predicates to be designed i.e AND  / OR / AND with NESTED OR e.t.c. This can be done right now
3. Searching by other data types

@jingtang10 
We can introduce this later 
[Reverse chaining](https://github.com/OpenSRP/fhircore/pull/57). Can be used when filtering patients with an observation / in a CarePlan


--
At the fhircore level we can add a way to generate the different search combinations using the UI. Also implementing the code that would

**Comment by rkodev** [comment]
@jingtang10 This issue is blocked by partial search. Delivering partial search, that's case insensitive,  on patient names should enable mark this issue as done

**Comment by jingtang10** [comment]
https://github.com/google/android-fhir/issues/376
https://github.com/google/android-fhir/issues/400

**Comment by jingtang10** [comment]
https://github.com/google/android-fhir/issues/401

**Comment by rkodev** [comment]
We will need to at least close search using OR parameters or resolving the bug on the index table reported here to fully resolve this issue https://github.com/google/android-fhir/issues/450

**Comment by f-odhiambo** [comment]
@rkodev Kindly advise if we should close this issue and create a new ticket for the outstanding work

**Comment by f-odhiambo** [comment]
**Update**
1. Count PR merged
2. Moving to OR operator
